File: src/main/java/akostenko/aicars/race/car/CarTelemetryScalar.java
package akostenko.aicars.race.car;

import org.newdawn.slick.Color;

public class CarTelemetryScalar {

    private final Color color;
    private final String name;
    private final int precision;
    private final String units;
    private final double value;
    private final String textValue;


    private CarTelemetryScalar(String name, double value, String units, int precision, Color color, String textValue) {
        this.color = color;
        this.name = name;
        this.precision = precision;
        this.units = units;
        this.value = value;
        this.textValue = textValue;
    }

    public CarTelemetryScalar(String name, double value, String units, int precision, Color color) {
        this(name, value, units, precision, color, null);
    }

    public CarTelemetryScalar(String name, double value, String units) {
        this(name, value, units, 0, CarTelemetry.textColor);
    }

    public CarTelemetryScalar(String name, String textValue) {
        this(name, 0, null, 0, CarTelemetry.textColor, textValue);
    }

    public Color color() {
        return color;
    }

    public String name() {
        return name;
    }

    public String value() {
        return textValue == null ? String.format("%." + precision + "f " + units, value) : textValue;
    }
}

File: src/main/java/akostenko/aicars/race/car/CarTelemetryScalar.kt
package akostenko.aicars.race.car

import org.newdawn.slick.Color

data class CarTelemetryScalar
private constructor(val name: String,
                    private val value: Double,
                    private val units: String?,
                    private val precision: Int,
                    val color: Color,
                    private val textValue: String?) {

    @JvmOverloads
    constructor(name: String,
                value: Double,
                units: String,
                precision: Int = 0,
                color: Color = CarTelemetry.textColor)
            : this(name, value, units, precision, color, null)

    constructor(name: String, textValue: String)
            : this(name, 0.0, null, 0, CarTelemetry.textColor, textValue)

    fun textValue(): String = textValue ?: String.format("%." + precision + "f " + units, value)
}
